### PR TITLE
add plugin support for paddle-trt

### PR DESCRIPTION
**Pull request by NHZlX**
## Paddle-TRT 添加Plugin

**NOTE:** 由于预测接口发生变化，在[这个pr](https://github.com/PaddlePaddle/Paddle/pull/13914)前的trt测试代码已经不能使用，请update到新的接口。

**Step one:** 添加plugin

```
参照 paddle/fluid/inference/tensorrt/plugin/split_op_plugin.h 的实现
```

**Step two** 添加从fluid op -> trt 的conveter

```
参照 paddle/fluid/inference/tensorrt/convert/split_op.cc
```

**Step three** 添加conveter 的utest

```
参照 paddle/fluid/inference/tensorrt/convert/test_split_op.cc
```

**Step four**  模型测试

```
1. /paddle/fluid/inference/analysis/passes/ir_analysis_compose_pass.cc#L48 添加 split op
2. USE_TRT_CONVERTER(split); 添加到paddle/fluid/inference/api/analysis_predictor.cc#L551

3. 找一个模型测试。 参考: /paddle/fluid/inference/tests/api/trt_models_tester.cc 
```

### CODE REVIEW

File: paddle/fluid/inference/tensorrt/engine.cc [hunk 1]
```diff
@@ -254,6 +254,12 @@ void TensorRTEngine::freshDeviceId() {
   cudaSetDevice(device_);
 }
 
+nvinfer1::IPluginLayer *TensorRTEngine::addPlugin(
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/engine.cc]: AddPlugin, follow google style in the underlying codebase.

File: paddle/fluid/inference/tensorrt/plugin/serialize.h [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+#pragma once
+
+#include <cassert>
+#include <cstring>
+#include <type_traits>
+#include <vector>
+
+template <typename T>
+inline void serialize_value(void** buffer, T const& value);
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/plugin/serialize.h]: SerializeValue

File: paddle/fluid/inference/tensorrt/plugin/serialize.h [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+template <typename T>
+inline void deserialize_value(void const** buffer, size_t* buffer_size,
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/plugin/serialize.h]: DeserializeValue

File: paddle/fluid/inference/tensorrt/plugin/serialize.h [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+template <typename T>
+struct Serializer<T, typename std::enable_if<std::is_arithmetic<T>::value ||
+                                             std::is_enum<T>::value ||
+                                             std::is_pod<T>::value>::type> {
+  static size_t serialized_size(T const& value) { return sizeof(T); }
+  static void serialize(void** buffer, T const& value) {
+    ::memcpy(*buffer, &value, sizeof(T));
+    reinterpret_cast<char*&>(*buffer) += sizeof(T);
+  }
+  static void deserialize(void const** buffer, size_t* buffer_size, T* value) {
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/plugin/serialize.h]: All should follow google C style.

File: paddle/fluid/inference/tensorrt/plugin/serialize.h [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    assert(*buffer_size >= sizeof(T));
+    ::memcpy(value, *buffer, sizeof(T));
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/plugin/serialize.h]: std::memcpy?

File: paddle/fluid/inference/tensorrt/plugin/serialize.h [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    reinterpret_cast<char const*&>(*buffer) += sizeof(T);
+    *buffer_size -= sizeof(T);
+  }
+};
+
+template <>
+struct Serializer<const char*> {
+  static size_t serialized_size(const char* value) { return strlen(value) + 1; }
+  static void serialize(void** buffer, const char* value) {
+    ::strcpy(static_cast<char*>(*buffer), value);
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/plugin/serialize.h]: `std::strcpy`  in <cstring> ?

File: paddle/fluid/inference/tensorrt/plugin/serialize.h [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+  return Serializer<T>::serialized_size(value);
+}
+
+template <typename T>
+inline void serialize_value(void** buffer, T const& value) {
+  return Serializer<T>::serialize(buffer, value);
+}
+
+template <typename T>
+inline void deserialize_value(void const** buffer, size_t* buffer_size,
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/plugin/serialize.h]: All need follow google C style.

File: paddle/fluid/inference/tensorrt/plugin/trt_plugin.cc [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/inference/tensorrt/plugin/trt_plugin.h"
+
+namespace paddle {
+namespace inference {
+namespace tensorrt {
+
+void PluginTensorRT::serializeBase(void*& buffer) {
```
Comment (Superjomn) [inline paddle/fluid/inference/tensorrt/plugin/trt_plugin.cc:21]: All function name need to follow google c style.

Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/plugin/trt_plugin.cc:21]: Many of the functions in this file are overridden of `nvinfer1::IPluginExt`.  The interface from `https://docs.nvidia.com/deeplearning/sdk/tensorrt-api/c_api/classnvinfer1_1_1_i_plugin_ext.html` seems not follow the google c stype. This document might not be changed for many code formats look ugly.

File: paddle/fluid/inference/tensorrt/plugin/serialize.h (hunk 7 above)
Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/plugin/serialize.h]: Done!

File: paddle/fluid/inference/tensorrt/plugin/serialize.h (hunk 6 above)
Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/plugin/serialize.h]: Done!

File: paddle/fluid/inference/tensorrt/plugin/serialize.h (hunk 5 above)
Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/plugin/serialize.h]: Done!

File: paddle/fluid/inference/tensorrt/plugin/serialize.h (hunk 4 above)
Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/plugin/serialize.h]: Done!

File: paddle/fluid/inference/tensorrt/plugin/serialize.h (hunk 3 above)
Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/plugin/serialize.h]: Done!

File: paddle/fluid/inference/tensorrt/plugin/serialize.h (hunk 2 above)
Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/plugin/serialize.h]: Done!

File: paddle/fluid/inference/tensorrt/engine.cc (hunk 1 above)
Comment (NHZlX) [reply paddle/fluid/inference/tensorrt/engine.cc]: Done!

**Review by Superjomn — APPROVED** [review]
LGTM

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**